BDD

tomcsmith1990 opened the pull request:
I'm making a pull request for this just so that I can put all of my comments with the code.

Commit:
Big commit, SpecFlow BDD style tests using NCrunch.

Commit:
Added CashMaths feature and steps.

Commit:
BDD Cleanup as I understand how it works better.

Mainly using the parameters for steps a lot more.

Commit:
Minor cleanup

Commit:
Removing . from step.

Commit:
Attempt to add nSpec project - so far tests aren't working.

Commit:
Adding ToString feature.

Commit:
R# refactor.

Commit:
Files left out of previous commit.

tomcsmith1990 commented on BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature:
```diff
@@ -0,0 +1,19 @@
+
+@mytag
+Scenario: New account has zero cash.
+	Given I want to create an account
```
This `Given` seems redundant - can it be missed out? It's equivalent to a test which doesn't require any arrangement.

SuperDrew replied:
It is redundant, you'll see in it's step that there's a TODO: Do I need to do anything here?

tomcsmith1990 commented on BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature:
```diff
@@ -0,0 +1,19 @@
+Scenario: New account can't take null cash.
+	Given I want to create an account
+	And I have null cash
+	Then Creating an account should throw an argument null exception
```
This test feels very dev-oriented, due to talking about `null` and exceptions. I'm not sure whether there would be a more user focused way of writing it (`Given I have an unknown amount of cash Then creating an account should give an error`?), or whether the test fits at all. Maybe BDD doesn't lend itself well to testing failure conditions?

SuperDrew replied:
I agree with it feeling dev-orientated.  I think you can test failure conditions as user stories with BDD but you're right this particular example is a bad one and may be more suitable at a lower level.

tomcsmith1990 commented:
I think there's a more general question of who are these tests aimed at? It's quite nice being able to read them in English, but if we (development people) are going to be the only ones using them, would we be better just writing tests in code?

tomcsmith1990 commented on BankingKataSpecs/AccountSpecs/Features/DepositCash.feature:
```diff
@@ -0,0 +1,10 @@
+
+@mytag
+Scenario: Depositing my cash
+	Given I create an account with 0 cash
```
Is there any good Resharper or other (support) for tests like these? e.g. if I had misspelled one of these conditions and want to correct it, do I have to do a find/replace on the string or can I do an automated refactoring?

SuperDrew replied:
You can bring up a list of existing steps to pick from but I haven't found any refactoring support for them.

tomcsmith1990 commented on BankingKataSpecs/AccountSpecs/Features/ToStringOverride.feature:
```diff
@@ -0,0 +1,9 @@
+	As a math idiot
+	I want to be able to get an informative ToStirng for an account.
+
+@mytag
```
What do these tags do?

SuperDrew replied:
They are converted to categories in NUnit so can be used for running parts of your test suite.  @ignore tag also ignores the feature or scenario.

You can also use them for SpecFlow Hooks, for example you can have a BeforeScenario tag and hook that mean for any scenario with that tag the hook is run (this is a brief summary of research online, I've never used hooks or tags).

People also use them for tagging bug reports numbers etc...  You can have multiple tags.

tomcsmith1990 commented on BankingKataNSpec/DebuggerShim.cs:
```diff
@@ -0,0 +1,44 @@
+
+        //assert that there aren't any failures
+        results.Failures().Count().should_be(0);
+    }
```
Specflow definitely seems a lot nicer.

SuperDrew replied:
Agreed

SuperDrew commented:
Good question, who are they aimed at.  There are a few purposes for these tests.  
- Executable specification.  This means that anybody can read the behavioural specification and execute them.  The link between specification and execution can show how much we've implemented and what is working.
- Show who?  Product managers, sprint review attendees, UX, research, anybody who is non-technical can understand this.  It thus can become a communication and collaboration tool.  Further to this is means people other than devs or testers can write them.  Whether we use this is up for debate.
- Link our user stories directly to executable code gives us a data point for our definition of done.
- Encourage us to write tests at a behavioural level.

We could achieve this without using something like SpecFlow but it seems to make sense to use a framework that assists us write valuable tests.  So even if only devs read the tests I still think it's useful to do although not as useful as if its used as a collaboration tool.

Commit:
Removed redundant steps.

Commit:
BDD style cleanup
Removed implementation specific scenario and details.
Attempted to make naming more user story focused.
